src/vos/tests/vts_pm.c

```diff
@@ -1487,6 +1487,193 @@ small_sgl(void **state)
+
+	/* Allocate memory for the scatter-gather list */
+	rc = daos_sgl_init(&sgl, 1);
+	assert_int_equal(rc, 0);
+
+	d_iov_set(&sgl.sg_iovs[0], (void *)first, rex.rx_nr);
+
+	vts_dtx_begin_ex(&oid, arg->ctx.tc_co_hdl, epoch++, 0, 3, &dth);
+
+	/* Write the first value */
+	rc = vos_obj_update_ex(arg->ctx.tc_co_hdl, oid,
+			       0 /* epoch comes from dth */, 0, 0, &dkey, 1,
+			       &iod, NULL, &sgl, dth);
+	if (rc != 0)
+		goto tx_end;
+
+	/* Punch the akey */
+	dth->dth_op_seq = 2;
+	rc = vos_obj_punch(arg->ctx.tc_co_hdl, oid,
+			   0 /* epoch comes from dth */, 0, 0, &dkey, 1, &akey,
+			   dth);
+	if (rc != 0)
+		goto tx_end;
+
+	/* Do next update */
+	dth->dth_op_seq = 3;
+	rex.rx_idx = rex.rx_nr;
+	rex.rx_nr = strlen(second);
+	d_iov_set(&sgl.sg_iovs[0], (void *)second, strlen(second));
+	rc = vos_obj_update_ex(arg->ctx.tc_co_hdl, oid,
+			       0 /* epoch comes from dth */, 0, 0, &dkey, 1,
+			       &iod, NULL, &sgl, dth);
+tx_end:
+	xid = dth->dth_xid;
+	vts_dtx_end(dth);
+	assert_int_equal(rc, 0);
+
+	rc = vos_dtx_commit(arg->ctx.tc_co_hdl, &xid, 1, NULL);
+	assert_int_equal(rc, 1);
+
+	/* Now read back original # of bytes */
+	rex.rx_idx = 0;
+	rex.rx_nr = strlen(expected);
+	memset(buf, 'x', sizeof(buf));
+	d_iov_set(&sgl.sg_iovs[0], (void *)buf, strlen(expected));
+	rc = vos_obj_fetch(arg->ctx.tc_co_hdl, oid, epoch++, 0, &dkey, 1,
+			   &iod, &sgl);
+	assert_int_equal(rc, 0);
+
+	assert_memory_equal(buf, expected, strlen(expected));
+
+	/* Now repeat the test for single value */
+	vts_key_gen(&akey_buf[0], arg->akey_size, false, arg);
+
+	iod.iod_type = DAOS_IOD_SINGLE;
+	iod.iod_size = strlen(first);
+	iod.iod_name = akey;
+	iod.iod_recxs = NULL;
+	iod.iod_nr = 1;
+
+	/* Allocate memory for the scatter-gather list */
+	rc = daos_sgl_init(&sgl, 1);
+	assert_int_equal(rc, 0);
+
+	d_iov_set(&sgl.sg_iovs[0], (void *)first, iod.iod_size);
+
+	vts_dtx_begin_ex(&oid, arg->ctx.tc_co_hdl, epoch++, 0, 2, &dth);
+
+	/* Write the first value */
+	rc = vos_obj_update_ex(arg->ctx.tc_co_hdl, oid,
+			       0 /* epoch comes from dth */, 0, 0, &dkey, 1,
+			       &iod, NULL, &sgl, dth);
+	if (rc != 0)
+		goto tx_end2;
+
+	/* Punch the akey */
+	dth->dth_op_seq = 2;
+	rc = vos_obj_punch(arg->ctx.tc_co_hdl, oid,
+			   0 /* epoch comes from dth */, 0, 0, &dkey, 1, &akey,
+			   dth);
+tx_end2:
+	xid = dth->dth_xid;
+	vts_dtx_end(dth);
+	assert_int_equal(rc, 0);
+
+	rc = vos_dtx_commit(arg->ctx.tc_co_hdl, &xid, 1, NULL);
+	assert_int_equal(rc, 1);
+
+	/* Now read back original # of bytes */
+	iod.iod_size = 0;
+	memset(buf, 'x', sizeof(buf));
+	d_iov_set(&sgl.sg_iovs[0], (void *)buf, sizeof(buf));
+	rc = vos_obj_fetch(arg->ctx.tc_co_hdl, oid, epoch++, 0, &dkey, 1,
+			   &iod, &sgl);
+	assert_int_equal(rc, 0);
+
+	assert_int_equal(iod.iod_size, 0);
+	assert_memory_equal(buf, sexpected, strlen(sexpected));
+
+	/** Let's simulate replay scenario where rebuild replays a minor epoch
+	 * punch and an update with the same major epoch
+	 */
```

Comment:
I do not think this can happen in the real world, because we should not rebuild the non-committed DTX.

Reply:
This isn't a case of rebuilding non committed DTX.   Rebuild replays punches using only a major epoch.

Reply:
Let me expand a bit.   Consider the following scenario

update at epoch 1
snapshot at epoch 2
punch at epoch 3.1
update at epoch 3.2

when rebuilding the snapshot at 2, rebuild will replay a punch with "3".   Internally, if we see a replay update, we use max minor epoch.  If we see a replay punch, we use max - 1.   This code is replaying that scenario where rebuild replays a "future" punch.